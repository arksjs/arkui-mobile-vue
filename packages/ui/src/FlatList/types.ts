import type { Noop } from '../helpers'
import type { OnRefreshing, OnScroll } from '../ScrollView/types'
import type {
  ScrollTo,
  ScrollToIndex,
  ScrollToEnd,
  OnVisibleItemsChange,
  VirtualListProps
} from '../VirtualList/types'

export type OnEndReached = (payload: { distanceFromEnd: number }) => void

export interface FlatListProps extends VirtualListProps {
  endReachedThreshold?: number
  enablePullRefresh?: boolean
  lowerLoading?: boolean
}

export interface FlatListEmits {
  onEndReached?: OnEndReached
  onRefreshing?: OnRefreshing
  onScroll?: OnScroll
  onVisibleItemsChange?: OnVisibleItemsChange
}

export interface FlatListRef {
  scrollTo: ScrollTo
  scrollToIndex: ScrollToIndex
  scrollToEnd: ScrollToEnd
  recordInteraction: Noop
}

export type {
  OnEndReached as FlatListOnEndReached,
  OnRefreshing as FlatListOnRefreshing,
  OnScroll as FlatListOnScroll,
  OnVisibleItemsChange as FlatListOnVisibleItemsChange
}
