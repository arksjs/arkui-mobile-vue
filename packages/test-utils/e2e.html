<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
