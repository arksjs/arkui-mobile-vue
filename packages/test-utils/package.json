{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
