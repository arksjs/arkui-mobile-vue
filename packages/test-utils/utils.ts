export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
