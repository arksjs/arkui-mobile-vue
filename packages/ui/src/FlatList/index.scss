@import '../style/var.scss';

.ak-flat-list {
  position: relative;
  display: block;
  width: 100%;

  &_header,
  &_footer {
    .ak-flat-list.horizontal & {
      display: inline-block;
      vertical-align: top;
    }
  }
}
