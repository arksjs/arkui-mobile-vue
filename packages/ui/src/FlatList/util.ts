export const getClasses = (horizontal: boolean) => [
  'ak-flat-list',
  { horizontal }
]
