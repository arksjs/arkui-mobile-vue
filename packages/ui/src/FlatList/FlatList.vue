<template>
  <ScrollView
    :class="classes"
    :scrollX="horizontal"
    :scrollY="!horizontal"
    :lowerThreshold="lowerThreshold"
    :enablePullDirections="enablePullDirections"
    @scroll="onScroll"
    @scrollToLower="onScrollToLower"
    @refreshing="onRefreshing"
    ref="scrollViewRef"
  >
    <div class="ak-flat-list_header" v-if="$slots.header">
      <slot name="header"></slot>
    </div>
    <VirtualList
      ref="virtualListRef"
      :ids="ids"
      :initialHorizontal="initialHorizontal"
      :initialWaterfallCount="initialWaterfallCount"
      :itemSize="itemSize"
      :preLoad="preLoad"
      :approvedItemVisibleScale="approvedItemVisibleScale"
      @visibleItemsChange="onVisibleItemsChange"
      @resize="onResize"
    >
      <template #default="item">
        <slot :id="item.id" :index="item.index"> </slot>
        <slot name="separator" :id="item.id" :index="item.index"></slot>
      </template>
    </VirtualList>
    <LoadMore
      class="ak-flat-list_indicator"
      loading
      :vertical="horizontal"
      v-if="lowerLoading && ids.length > 0"
    >
      {{ locale.flatListLoadingText }}
    </LoadMore>
    <div class="ak-flat-list_empty" v-if="ids.length === 0">
      <slot name="empty"></slot>
    </div>
    <div class="ak-flat-list_footer" v-if="$slots.footer">
      <slot name="footer"></slot>
    </div>
  </ScrollView>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, shallowRef } from 'vue'
import { LoadMore } from '../LoadMore'
import { ScrollView } from '../ScrollView'
import { VirtualList } from '../VirtualList'
import type {
  OnVisibleItemsChange,
  ScrollToEnd,
  ScrollToIndex,
  ScrollTo,
  VirtualListRef
} from '../VirtualList/types'
import { useLocale } from '../ConfigProvider/context'
import type { OnRefreshing, OnScroll, PullDirection } from '../ScrollView/types'
import {
  emitRefreshingValidator,
  emitScrollValidator
} from '../ScrollView/props'
import {
  emitVisibleItemsChangeValidator,
  virtualListProps
} from '../VirtualList/props'
import { getClasses } from './util'
import { isNumber } from '../helpers/util'
import type { PropsToEmits } from '../helpers/types'
import type { FlatListEmits } from './types'

export default defineComponent({
  name: 'ak-flat-list',
  components: { ScrollView, LoadMore, VirtualList },
  props: {
    ...virtualListProps,
    // 决定当距离内容最底部还有多远时触发endReached回调
    endReachedThreshold: {
      type: Number,
      default: 0.5
    },
    // 是否开启下拉刷新
    enablePullRefresh: {
      type: Boolean,
      default: false
    },
    // 展示加载更多图标
    lowerLoading: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    endReached: payload => payload && isNumber(payload.distanceFromEnd),
    scroll: emitScrollValidator,
    refreshing: emitRefreshingValidator,
    visibleItemsChange: emitVisibleItemsChangeValidator
  } as PropsToEmits<FlatListEmits>,
  setup(props, { emit }) {
    const { locale } = useLocale()
    const scrollViewRef = shallowRef<InstanceType<typeof ScrollView> | null>(
      null
    )
    const virtualListRef = shallowRef<VirtualListRef | null>(null)

    const wrapperSize = ref(0)
    let horizontal = false

    if (props.initialWaterfallCount <= 1 && props.initialHorizontal) {
      horizontal = true
    }

    function getScrollContainer() {
      return scrollViewRef.value?.$el as HTMLElement
    }

    function onResize(size: number) {
      wrapperSize.value = size
    }

    const onRefreshing: OnRefreshing = (res, loadComplete) => {
      emit('refreshing', res, loadComplete)
    }

    const onVisibleItemsChange: OnVisibleItemsChange = res => {
      emit('visibleItemsChange', res)
    }

    /**
     * 主动通知列表发生了一个事件，以使列表重新计算可视区域
     */
    function recordInteraction() {
      virtualListRef.value?.recordInteraction()
    }

    const onScroll: OnScroll = res => {
      // scroll 事件回调
      emit('scroll', res)
    }

    function onScrollToLower() {
      const $el = getScrollContainer()
      const distanceFromEnd = horizontal
        ? $el.scrollWidth - $el.scrollLeft - $el.offsetWidth
        : $el.scrollHeight - $el.scrollTop - $el.offsetHeight

      emit('endReached', {
        distanceFromEnd
      })
    }

    const scrollTo: ScrollTo = options => {
      virtualListRef.value?.scrollTo(options as any)
    }
    const scrollToIndex: ScrollToIndex = options => {
      virtualListRef.value?.scrollToIndex(options as any)
    }
    const scrollToEnd: ScrollToEnd = animated => {
      virtualListRef.value?.scrollToEnd(animated)
    }

    const lowerThreshold = computed(() => {
      return wrapperSize.value * props.endReachedThreshold
    })

    const enablePullDirections = computed<PullDirection[]>(() => {
      if (props.enablePullRefresh) {
        return horizontal ? ['right'] : ['down']
      }

      return []
    })

    onMounted(() => {
      if (scrollViewRef.value) {
        virtualListRef.value?.resetScrollContainer(getScrollContainer())
      }
    })

    const classes = getClasses(horizontal)

    return {
      classes,
      scrollViewRef,
      virtualListRef,
      horizontal,
      lowerThreshold,
      enablePullDirections,
      onRefreshing,
      recordInteraction,
      onScroll,
      onScrollToLower,
      scrollTo,
      scrollToIndex,
      scrollToEnd,
      locale,
      onVisibleItemsChange,
      onResize
    }
  }
})
</script>
