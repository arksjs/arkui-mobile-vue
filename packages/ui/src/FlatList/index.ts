import { withInstall } from '../helpers/with-install'
import FlatList from './FlatList.vue'

export { FlatList }
export const install = withInstall(FlatList)
export default FlatList
