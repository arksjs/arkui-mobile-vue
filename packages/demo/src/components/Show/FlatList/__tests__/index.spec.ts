import { shallowMount } from '@vue/test-utils'
import { FlatList } from '@/FlatList'

const data: {
  id: number
  text: string
}[] = []

for (let i = 0; i < 10; i++) {
  data.push({
    id: i + 1,
    text: `Item ${i + 1}.`
  })
}

describe('FlatList', () => {
  test('should render default slot correctly', () => {
    const wrapper = shallowMount(FlatList, {
      props: {
        data
      },
      scopedSlots: {
        default: `<div>{{ props.item.text }}</div>`
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
