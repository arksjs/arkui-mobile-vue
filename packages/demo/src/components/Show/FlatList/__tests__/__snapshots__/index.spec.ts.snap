// Vitest Snapshot v1

exports[`FlatList > should render default slot correctly 1`] = `"<scroll-view-stub scrollx=\\"false\\" scrolly=\\"true\\" scrollanimated=\\"false\\" upperthreshold=\\"50\\" lowerthreshold=\\"0\\" enablepulldirections=\\"\\" pullrefreshthreshold=\\"48\\" class=\\"fx-flat-list\\" data=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\"></scroll-view-stub>"`;
